pkg/divert/istio/divert.go

```diff
@@ -0,0 +1,212 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package istio
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/okteto/okteto/pkg/k8s/virtualservices"
+	oktetoLog "github.com/okteto/okteto/pkg/log"
+	"github.com/okteto/okteto/pkg/model"
+	istioV1beta1 "istio.io/client-go/pkg/apis/networking/v1beta1"
+	istioclientset "istio.io/client-go/pkg/clientset/versioned"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	UPDATE_CONFLICT_RETRIES = 10
+)
+
+// Driver weaver struct for the divert driver
+type Driver struct {
+	Client      kubernetes.Interface
+	IstioClient istioclientset.Interface
+	Manifest    *model.Manifest
+	cache       *divertIstioCache
+}
+
+type divertIstioCache struct {
+	divertVirtualServices    map[string]*istioV1beta1.VirtualService
+	developerVirtualServices map[string]*istioV1beta1.VirtualService
+}
+
+func (d *Driver) Deploy(ctx context.Context) error {
+	for name := range d.cache.divertVirtualServices {
+		select {
+		case <-ctx.Done():
+			oktetoLog.Infof("deployDivert context cancelled")
+			return ctx.Err()
+		default:
+			oktetoLog.Spinner(fmt.Sprintf("Diverting virtual service %s/%s ...", d.Manifest.Deploy.Divert.Namespace, name))
+			if err := d.retryTranslateDivertVirtualService(ctx, name); err != nil {
+				return err
+			}
+
+			if !isInDivertVirtualServices(name, d.Manifest.Deploy.Divert.VirtualServices) {
+				continue
+			}
+			if err := d.retryTranslateDeveloperVirtualService(ctx, name); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (d *Driver) Destroy(ctx context.Context) error {
+	for name := range d.cache.divertVirtualServices {
+		select {
+		case <-ctx.Done():
+			oktetoLog.Infof("destroyDivert context cancelled")
+			return ctx.Err()
+		default:
+			oktetoLog.Spinner(fmt.Sprintf("Destroying divert un virtual service %s/%s ...", d.Manifest.Deploy.Divert.Namespace, name))
+			if err := d.retryRestoreDivertVirtualService(ctx, name); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (d *Driver) GetDivertNamespace() string {
+	if d.Manifest.Deploy.Divert.Namespace == d.Manifest.Namespace {
+		return ""
+	}
+	return d.Manifest.Deploy.Divert.Namespace
+}
+
+func (d *Driver) InitCache(ctx context.Context) error {
+	d.cache = &divertIstioCache{
+		divertVirtualServices:    map[string]*istioV1beta1.VirtualService{},
+		developerVirtualServices: map[string]*istioV1beta1.VirtualService{},
+	}
+	vsList, err := virtualservices.List(ctx, d.Manifest.Deploy.Divert.Namespace, d.IstioClient)
+	if err != nil {
+		return err
+	}
+	for i := range vsList {
+		d.cache.divertVirtualServices[vsList[i].Name] = vsList[i]
+	}
+
+	vsList, err = virtualservices.List(ctx, d.Manifest.Namespace, d.IstioClient)
```

Comment:
could we use a different variable here in order to not mix them up? something like `devVsList`